requests-for-comments/2022-06-01-session-analysis.md

```diff
@@ -0,0 +1,118 @@
+
+### 4. Session length distributions
+
+Proposal: We add session length as an option in the breakdown properties. Additionally, we allow users to “auto-bucket” breakdowns that are on numeric properties. 
+
+![image](https://user-images.githubusercontent.com/4813045/171521347-8f3fd16c-28ab-49ac-9baa-c02097d4c42c.png)
+
+Details: The “auto-bucket” functionality would require more work to figure out exactly how this should work + what the UX should be.
+
+### 5. Count of a specific event per session over time
+
+Proposal: Users accomplish this using formulas. They would make a series that is (a) Session count and (b) total count, and then divide (b) by (a).
+
+![image](https://user-images.githubusercontent.com/4813045/171521365-2b353bb0-7634-45fa-84e3-1463eb351fc0.png)
+
+### 6. Limit a funnel conversion to a single session
+
+Proposal: In funnels, we add “Unique sessions” as an aggregation option. Setting this value would enforce that the sequence of events must occur within a specific session.
+
+![image](https://user-images.githubusercontent.com/4813045/171521401-3b7a234c-24b4-45cd-a3d0-15ab4243efb0.png)
+
+
+### 7. Limit a path conversion to a single session
+
+Proposal: I’d consider this to be a lower priority, but we should update the logic in paths to respect the concept of a session. Today, it uses a 30 min gap between events to split paths - which is essentially the same as sessions, but subtly different.
+
+One issue that we will have to figure out is being backward compatible to data that was generated prior to session support.
+
+## Work breakout:
+
+To accomplish the above, we would need to do the following work:
+
+- Trends:
+    - Add “unique sessions” as an event aggregation option in trends
+    - Add the concept of “session properties”, and specifically the “session length” property to the taxonomic filter.
+    - Allow for session properties to be used as a property aggregation in trends.
+    - Allow for session properties to be used in breakdowns
+    - Add the ability to “auto bucket” when breaking down by numeric properties
+    - Note: while not included in the requirements above, it’d be cool to allow filtering based on session length everywhere. (e.g. total count of pageviews where the corresponding session length > 1 min)
+- Funnels
+    - Add sessions as a aggregation target
+- Paths
+    - Use sessions instead of the 30 min gap (need to consider backward compatibility here)
+
+## Questions + notes
+
+1. Do we need a sessions table?
+    1. If we add session length as a property option, those queries will have to generate a list of session durations on the fly. This query can’t be limited to specific event types because it has to query across all events to ensure it finds the first and last event in the session. As a result it might be a hefty query. To optimize this, we could create a sessions table in clickhouse (probably using AggregatingMergeTree) ****that contains the session_id and it’s length
```

Comment:
I used to do this madness after calcluating the 30minute window so with session_id alone, the query should actually be ok

[Basic query example](https://metabase.posthog.net/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJTRUxFQ1QgZGF0ZV9kaWZmKCdzZWNvbmQnLCBtaW5fdGltZSwgbWF4X3RpbWUpIGFzIHNlY29uZHMsIG1heCh0aW1lc3RhbXApIG1heF90aW1lLCBtaW4odGltZXN0YW1wKSBtaW5fdGltZSBGUk9NIGV2ZW50cyB3aGVyZSB0ZWFtX2lkID0gMiBhbmQgTk9UICRzZXNzaW9uX2lkID0gJycgQU5EIHRpbWVzdGFtcCA-ICcyMDIyLTA1LTAxIDIzOjU5OjU5JyBHUk9VUCBCWSAkc2Vzc2lvbl9pZCIsInRlbXBsYXRlLXRhZ3MiOnt9fSwiZGF0YWJhc2UiOjJ9LCJkaXNwbGF5IjoidGFibGUiLCJ2aXN1YWxpemF0aW9uX3NldHRpbmdzIjp7fX0=)

Reply:
not bad!